File: simulator/abilities/on_hit_prone.py
from ..abilities.on_hit_effect import OnHit
from ..misc import roll_saving_throw, reconcile_roll_types
from ..conditions import Conditions, Condition, apply_condition
import logging

logger = logging.getLogger("Encounterra")


class OnHitProne(OnHit):
    def __init__(self, st, dc, name="On Hit Prone"):
        self.st = st
        self.dc = dc
        self.name = name

    def hit(self, attacker, attack, target, multiplier, dmg_so_far):
        saved = roll_saving_throw(target.saving_throws[self.st], self.dc, reconcile_roll_types(target.saving_throws_roll_type_mod[self.st]))
        if not saved:
            logger.info(f"{target} is knocked prone")
            apply_condition(target, Condition(Conditions.PRONE, attacker))
        return None

    def calculate_threat(self, attacker, target, **kwargs):
        return 0  # TODO

File: c++/abilities/on_hit_prone.cpp
#include "abilities/on_hit_prone.hpp"

#include <iostream>
#include "core/combatant.hpp"

namespace enc
{
  std::vector<std::pair<int, DamageType>>
  OnHitProne::hit(Combatant *attacker, Actoid * /*attack*/, Combatant *target, double /*multiplier*/, double /*dmgSoFar*/)
  {
    // Already prone: nothing to do.
    if(target->isAffectedBy(Conditions::PRONE))
      {
        return {};
      }

    if(!_requiresSave)
      {
        // Automatic prone (e.g. 2024 Dire Wolf Bite): no saving throw, but only if the target is small enough.
        if(target->getSize() > _maxSize)
          {
            return {};
          }
        std::cout << target->_name << " is knocked Prone (no save)" << std::endl;
        target->applyCondition(Condition(Conditions::PRONE, attacker, nullptr, target));
        return {};
      }

    bool saved = rollSavingThrow(target->getSavingThrow(_saveType), _dc, RollType::STRAIGHT);
    if(!saved)
      {
        std::cout << target->_name << " is knocked Prone (DC " << _dc << ")" << std::endl;
        target->applyCondition(Condition(Conditions::PRONE, attacker, nullptr, target));
      }
    return {};
  }
}
